### ISSUE TITLE
v2.x : Breaks `assert_hostname = False`

### ISSUE THREAD

**Issue by Matthew1471** [urllib3/urllib3#3051]
Hi,

If I upgrade to urllib3 v2.x then my disabling of hostname checking code (based upon `assert_hostname = False` as per https://github.com/urllib3/urllib3/pull/194 and https://github.com/psf/requests/pull/1405) no longer works in v2.x.

I have read that v2.x now does SAN instead of CN checking (https://urllib3.readthedocs.io/en/stable/v2-migration-guide.html#stop-verifying-commonname-in-certificates) but I don't believe the `assert_hostname = False` functionality should have also been removed / silently ignored as part of it (that param looks like it could now work with SAN hostnames instead?).

I am creating a HTTPAdapter via Requests:

```python
# Requests references passed from calling code.
from requests.adapters import HTTPAdapter, DEFAULT_POOLBLOCK

class IgnoreHostnameAdapter(HTTPAdapter):

    def init_poolmanager(self, connections, maxsize, block=DEFAULT_POOLBLOCK, **pool_kwargs):
        pool_kwargs['assert_hostname'] = False
        super(IgnoreHostnameAdapter, self).init_poolmanager(connections=connections, maxsize=maxsize, block=block, **pool_kwargs)
```

which is mounted to the session elsehwere:

```python
# Make the session verify all HTTPS requests with trust for this certficiate.
self.session.verify = 'configuration/gateway.cer'

# Requests to this host will ignore the hostname in the certificate being incorrect.
self.session.mount(self.host, IgnoreHostnameAdapter())
```

This works in v1.x (`pip install urllib3==1.26.16`) but breaks in v2.x releases (`pip install urllib3 --upgrade`).

Is there an equivalent "`subjectAltName` hostname check" disable feature in v2.x?

This is for connecting to an embedded device where the certificate details are self-signed with an incorrect hostname and I would like to pin the cert rather than disable certificate verification entirely.

Using requests' `verify=False` to disable all certificate verification of course works but I would like to continue to use requests' `session.verify` to pin the certificate (just ignore the mismatching hostname).

Thank you for your time.

**Comment by pquentin** [comment]
Hello, `assert_hostname=False` is definitely supposed to work: we even have an unit test for it. I'll try to reproduce locally.

**Comment by sg3-141-592** [comment]
I've been having a look this morning. I've got a clean recreation adding this additional test to `test_https.py` which is passing with `urllib3==1.26.16` and failing on `urllib3==2.0.2`. 

``` python
class TestHTTPS_Hostname:
    def test_assert_hostname_invalid_san(self, no_localhost_san_server: ServerConfig) -> None:
        """Ensure SAN errors are not detected while assert_hostname is false"""
        with HTTPSConnectionPool(
            no_localhost_san_server.host,
            no_localhost_san_server.port,
            cert_reqs="CERT_REQUIRED",
            ca_certs=no_localhost_san_server.ca_certs,
            assert_hostname=False
        ) as https_pool:
            https_pool.request("GET", "/")
```

Just following with the debugger now to see how the `assert_hostname` param is getting handled differently in the new version.

**Comment by sg3-141-592** [comment]
Some work in progress, it's coming down to this function in `ssl.py` `SSLSocket.do_handshake()`. It is not throwing in v1.26.16 but is in v2.0.2. Just looking for what's changed in the `SSLSocket` that it's throwing now.
 
<img width="553" alt="image" src="https://github.com/urllib3/urllib3/assets/5122866/a6f3c174-8079-4d6b-bce0-c74a790a559b">

I can see in the Python docs [https://docs.python.org/3/library/ssl.html#ssl.SSLSocket.do_handshake](https://docs.python.org/3/library/ssl.html#ssl.SSLSocket.do_handshake) this comment that looks promising

```
Changed in version 3.7: Hostname or IP address is matched by OpenSSL during handshake. The function [match_hostname()](https://docs.python.org/3/library/ssl.html#ssl.match_hostname) is no longer used. In case OpenSSL refuses a hostname or IP address, the handshake is aborted early and a TLS alert message is sent to the peer.
```

**Comment by pquentin** [comment]
Fixed in #3055.

**Comment by Matthew1471** [comment]
Just wanted to say thank you :)
